test/tenant_token_test.dart

```diff
@@ -0,0 +1,120 @@
+import 'package:dart_jsonwebtoken/dart_jsonwebtoken.dart';
+import 'package:meilisearch/meilisearch.dart';
+import 'package:meilisearch/src/tenant_token.dart';
+import 'package:test/test.dart';
+
+import 'utils/books.dart';
+import 'utils/client.dart';
+
+void main() {
+  final Map<String, dynamic> _searchRules = {"*": null};
+
+  group('Tenant Tokens', () {
+    setUpClient();
+
+    final List<dynamic> possibleRules = [
+      {'*': {}},
+      {'*': null},
+      ['*'],
+      {
+        '*': {"filter": 'tag = Tale'}
+      },
+      {"books": {}},
+      {"books": null},
+      ['books'],
+      {
+        "books": {"filter": 'tag = comedy AND book_id = 1'}
+      }
+    ];
+
+    group('client.generateTenantToken', () {
+      test('decodes successfully using apiKey from instance', () {
+        final token = client.generateTenantToken(_searchRules);
+
+        expect(() => JWT.verify(token, SecretKey(client.apiKey!)),
+            returnsNormally);
+      });
+
+      test('decodes successfully using apiKey from param', () {
+        final key = sha1RandomString();
+        final token = client.generateTenantToken(_searchRules, apiKey: key);
+
+        expect(() => JWT.verify(token, SecretKey(key)), returnsNormally);
+      });
+
+      test('throws InvalidApiKeyException if all given keys are invalid', () {
+        final custom = MeiliSearchClient(testServer, null);
+
+        expect(() => custom.generateTenantToken(_searchRules),
+            throwsA(isA<InvalidApiKeyException>()));
+      });
+
+      test('invokes search successfully with the new token', () async {
+        final admKey = await client.createKey(indexes: ["*"], actions: ["*"]);
+        final admClient = MeiliSearchClient(testServer, admKey.key);
+        await createBooksIndex(uid: 'books');
+        await admClient
+            .index('books')
+            .updateFilterableAttributes(['tag', 'book_id']).waitFor();
+
+        possibleRules.forEach((data) async {
+          final token = admClient.generateTenantToken(data);
+          final custom = MeiliSearchClient(testServer, token);
+
+          expect(() async => await custom.index('books').search(''),
+              returnsNormally);
+        });
+      });
+    });
+
+    group('tenant_token.generateToken', () {
+      test('generates a signed token with given key', () {
+        final key = sha1RandomString();
+        final token = generateToken(_searchRules, key);
+
+        expect(() => JWT.verify(token, SecretKey(key)), returnsNormally);
+        expect(() => JWT.verify(token, SecretKey('not-the-same-key')),
+            throwsA(isA<JWTError>()));
+      });
+
+      test('does not generate a signed token without a key', () {
+        expect(() => generateToken(_searchRules, ''),
+            throwsA(isA<InvalidApiKeyException>()));
+      });
+
+      test('generates a signed token with a given expiration', () {
```

Comment:
Maybe it could be great to add a real test (on Meilsearch) with a expires at? We can't know for sure the format of the timestamp is good and if someday we made some modifications to it.

Reply:
Yes thats a good idea! 🌮 

But, since the tenant token feature, aim to generate a token complaint with the JWT spec our current implementation around the `exp` is ok with the spec. So I think there is no need to add a new validation around this information :)

From the RFC: https://www.rfc-editor.org/rfc/rfc7519#section-4.1.4

> **"exp" (Expiration Time) Claim**
 The "exp" (expiration time) claim identifies the expiration time on
   or after which the JWT MUST NOT be accepted for processing.  The
   processing of the "exp" claim requires that the current date/time
   MUST be before the expiration date/time listed in the "exp" claim.
   Implementers MAY provide for some small leeway, usually no more than
   a few minutes, to account for clock skew.  Its value MUST be a number
   containing a **_NumericDate_** value.  Use of this claim is OPTIONAL.


> **NumericDate**
      A JSON numeric value representing the number of seconds from
      1970-01-01T00:00:00Z UTC until the specified UTC date/time,
      ignoring leap seconds.  This is equivalent to the IEEE Std 1003.1,
      2013 Edition [[POSIX.1](https://www.rfc-editor.org/rfc/rfc7519#ref-POSIX.1)] definition "Seconds Since the Epoch", in
      which each day is accounted for by exactly 86400 seconds, other
      than that non-integer values can be represented.  See [RFC 3339](https://www.rfc-editor.org/rfc/rfc3339)
      [[RFC3339](https://www.rfc-editor.org/rfc/rfc3339)] for details regarding date/times in general and UTC in
      particular.

Reply:
Yes, but as long as Meilisearch is not yet stable I think it is always good to create real tests. I find it safer, it allows being sure where the problem comes from if something is modified. I think that when this day will come it will be good to have mocks to test only SDK. But it's like if you prefer there are no obligations.